### ISSUE TITLE
Flatlist not re-rendering on update/delete but does on adding

### ISSUE THREAD

**Issue by mahomahoxd** [mobxjs/mobx#1236]
I have an issue with my FlatList implementation. I'm facing issues regarding updating the FlatList when the data inside it has changed. Mainly the arrays are changed, its just not updating the Flatlist until I manually call "this.setState({})", after that it seems to refresh everything nicely. My question was, if there is a neater solution for this as I can imagine, modifying the state like this might have side-effects.

I have tried the approach of having observer(({item}) => instead of just the handler, this unfortunately gave me the error of calling a function while its a class. I've also tried to put the data of the FlatList to toJS(thearray) but that did not work either. I have also tried the component of <Observer> but that had the same effect of not changing anything. I think I might be doing something wrong in mutating the array but I am not entirely sure how or what I am doing wrong. When adding, it's all good, but as soon as I delete or update items in the list, it does not re-render. 

This issue might be duplicate, although I cannot find the case I am in, more like similar cases that suggest me the things I have done but don't work for me.

The MobX version I am using is 3.3.0 with MobX-React that is 4.3.3.

This is the store I am using : 

```
 export class TodoStore {
     @observable todosFinished = [];
     @observable todosUnfinished = [];

     constructor(todoItems){
      this.todosUnfinished.replace(todoItems.filter(todo => todo.done === false));
      this.todosFinished.replace(todoItems.filter(todo => todo.done === true));
     }
     @action addTodo(todo) {
              (todo.done === true) ? this.todosFinished.push(todo) : this.todosUnfinished.push(todo);
      }
     @action updateTodo(todo, index) {
         if(todo.done === true) { 
               this.todosUnfinished.splice(index,1);
               this.todosFinished.push(todo);
         } else {
              this.todosFinished.splice(index,); 
              this.todosUnfinished.push(todo);
      } 
 
      @action removeTodo(todo, index) {
             (todo.done === true) ? this.todosFinished.remove(todo) : this.todosUnfinished.remove(todo);
      }
```

This is a rough implementation of what I have on the flatlist 
```
           <FlatList
             style={{ flex: 1 }}
             data={toJS(todoStore.unfinishedItems)}
             extraData={todoList.unfinishedItems.length}
             renderItem={this.renderFlatListItem}
             keyExtractor={this.keyExtractor}
           />
```

And the renderFlatListItem function 

```
 
renderFlatListItem = ({ item, index }) => (
           (this.props.readOnly)
                ? <ROTodoItem todo={item} />
               : <ETodoItem
                         todo={item}
                         onPress={() => this.props.onPressTodoItem(item, index)} // This will update the .done state with the updateTodo function from the store
                         onDeleteTodo={() => {
                         this.deleteTodo(item); // This will call the removeTodo from the store
                        // Temporary fix
                        this.setState({}); // This is the thing that works for me, but it feels like a dirty/temporary fix
                 }}
            />);
```
If anyone could assist me with this issue, I'll be happy. Thanks in advance!

**Comment by urugator** [comment]
Can't spot the problem, but is there a reason to not to use computed instead of manually synchronizing 3 arrays?
```javascript
// considering the "todo.done" is observable as well
@observable todoItems = [];
@computed get finished() {
  return this.todoItems.filter(todo => todo.done)
}
@computed get unfinished() {
  return this.todoItems.filter(todo => !todo.done)
}
```
Note that FlatList must be inside observer (or to be precise - the `toJS(todoStore.unfinishedItems)` must be called in observer and FlatList must be it's child).

**Comment by mahomahoxd** [comment]
Well yes, the reason for the multiple lists is more or less because FlatList had some issues re-rendering the entire list as far as I can remember. It could be that I did that wrong too. I don't get quite what you mean with the Note, do you mind to elaborate on it? I think I might have it similarly but I'm not sure

**Comment by urugator** [comment]
> elaborate on it

The FlatList itself isn't reactive, it has no notion of Mobx. 
FlatList simply re-renders, when it's parent component re-renders and passes new data via prop to FlatList. 
So you can forget about FlatList and just try to make sure that the parent component is re-rendered whenever the relevant array is modified. To do so the parent component must be `observer`:
```javascript
const Parent = observer({ todoStore } => {
   // the observer reacts only to things which are accessed inside observer
   // we are going to pass the array to 3rd party component (FlatList), which is not an observer
   // so to make this component react to array and everything inside
   // we have to access the array and everything inside
   // to do so we can use toJS, which creates a deep copy by traversing the array (acceesing everything)
   const deepCopy = toJS(todoStore.todosUnfinished)
   // at this point the component is subscribed to `todoStore.todosUnfinished`   
})
// or
@observer class Parent extends React.Component {/*...*/}
```
I also think that you can remove `extraData` prop ... when you use `toJS` a new array is created and passed every time, therefore the "pure render check" should fail and force FlatList to re-render.

Also note that `toJS` creates a deep copy, therefore the `todo` item recieved in `update/remove` is different than the one in the array. Thats probably the reason why it doesn't work.

**Comment by urugator** [comment]
If you want to subscribe for everything inside the array and maintain the original references, you may try:
```javascript
<FlatList
             style={{ flex: 1 }}
             data={todoStore.todosUnfinished} // pass original array to maintain references            
             extraData={toJS(todoStore.todosUnfinished)} // call toJS to subscribe and pass copy to invalidate FlatList
             renderItem={this.renderFlatListItem}
             keyExtractor={this.keyExtractor}
           />
```

**Comment by mahomahoxd** [comment]
I am not sure, but I think it started to work when I put @observer above the TodoList class, for some reason it didn't work earlier today when I did the same thing but suddenly things started to work. I also have adjusted the 2 lists into one so that its easier to maintain, in the exact same way as you described. I've learned at least that the parent needs to change props before the FlatList will rerender, I never actually knew. In any case thank you for your effort! I'll close the issue now 👍
